### ISSUE TITLE
Elasticsearch like alternative 

### ISSUE THREAD

**Issue by takov751** [mastodon/mastodon#20743]
### Pitch

considering taking a look at https://github.com/meilisearch/meilisearch 
as elasticsearch alternative. Im by no mean know if its viable . Just wanted to put it out there so you might find it useful

### Motivation

In case someone encounters bugs or just unhappy with elasticsearch . Would be really nice to have another solution.

**Comment by LeoWinterDE** [comment]
Just use [opensearch](https://opensearch.org/). Rel #18535

**Comment by berkes** [comment]
Meilisearch is really a different model than OpenSearch or Elasticsearch and offers a lot of benefits over Elasticsearch. 

1. It's far simpler to host. It's a single binary that is executed, nothing more. Elastic/OpenSearch is a monster to host properly with all the clustering and balancing going on.
2. It's  very lightweight.
3. For smaller or simpler datasets, it performs better than elasticsearch given the same resources. Mastodon's search is small and simple.
4. Backups/restores are very easy.

**Comment by takov751** [comment]
That was my initial thoughts when i decided i will mention it here. It does have it's place, however it needs to be tested in the wild

**Comment by Exagone313** [comment]
Note that there are similar projects that exist, such as [Sonic](https://github.com/valeriansaliou/sonic). My use-case of any of the choices would be to have a tool that has a low footprint on the server, compared to ElasticSearch and its forks.

I'll probably take a look on how difficult the implementation would be in Mastodon.

**Comment by berkes** [comment]
It sounds as if a common pattern: ports and adapters might work here. Rails uses this a lot. 

For this to work, an interface must be designed (Ruby has no interfaces, unfortunately, so it needs to live as documentation) and then implemented for PostgreSQL and Elastic. Once that is in place, any third party can release gems, or incorporate in core.

For Mastodon, this approach would mean that nothing functional changes, which probably makes it easier to review and merge.

**Comment by TheEssem** [comment]
Sonic was discussed before, it may not be suitable for Mastodon's use cases: https://github.com/mastodon/mastodon/issues/17068#issuecomment-980782635

**Comment by eliastorres** [comment]
An alternative could be Manticore https://github.com/manticoresoftware/manticoresearch/ written in C++ uses low RAM unlike Elastic Search that eats lots of memory
